src/main/protobuf/remote_output_service.proto

```diff
@@ -0,0 +1,320 @@
+  rpc FinalizeBuild(FinalizeBuildRequest) returns (google.protobuf.Empty);
+}
+
+message CleanRequest {
+  // The output base identifier that was provided to
+  // StartBuildRequest.output_base_id whose data needs to be removed.
+  string output_base_id = 1;
+}
+
+message StartBuildRequest {
+  // A client-chosen value that uniquely identifies the workspace for
+  // which the build is being started.  This value must be set to ensure
+  // that the remote output service is capable of managing builds for
+  // distinct workspaces concurrently.
+  //
+  // This value must be a valid filename for the operating system on
+  // which the remote output service and client are being executed. This
+  // allows the remote output service to create one subdirectory per
+  // project that needs to be built.
+  //
+  // By default, Bazel sets this value to the MD5 sum of the absolute
+  // path of the workspace directory. This is generally sufficient,
+  // though a more complex scheme may necessary in case the file system
+  // namespace is virtualized.
+  //
+  // Starting a build finalizes any previous build with the same
+  // output_base_id that has not been finalized yet.
+  string output_base_id = 1;
+
+  // A client-chosen value that uniquely identifies this build. This
+  // value must be provided to most other methods to ensure that
+  // operations are targeted against the right output path.
+  string build_id = 2;
+
+  // The instance name that the client uses when communicating with the
+  // remote execution system. The remote output service uses this value
+  // when loading objects from the Content Addressable Storage.
+  string instance_name = 3;
+
+  // The digest function that the client uses when communicating with
+  // the remote execution system. The remote output service uses this
+  // value to ensure that FileStatus responses contain digests that were
+  // computed with right digest function.
+  build.bazel.remote.execution.v2.DigestFunction.Value digest_function = 4;
+
+  // The absolute path at which the remote output service exposes its
+  // output paths, as seen from the perspective of the client.
+  //
+  // This value needs to be provided by the client, because file system
+  // namespace virtualization may cause this directory to appear at a
+  // location that differs from the one used by the service.
+  //
+  // The purpose of this field is to ensure that the remote output
+  // service is capable of expanding symbolic links containing absolute
+  // paths.
+  string output_path_prefix = 5;
+
+  // A map of paths on the system that will become symbolic links
+  // pointing to locations inside the output path. Similar to
+  // output_path_prefix, this option is used to ensure the remote output
+  // service is capable of expanding symbolic links.
+  //
+  // Map keys are absolute paths, while map values are paths that are
+  // relative to the output path.
+  map<string, string> output_path_aliases = 6;
```

Comment:
What is the purpose of that the FUSE system? Especially given we fix up all the paths to not have the `bazel-out` in them? Why is that a map?

Reply:
The `BatchStat()` call has a `follow_symlinks` option. This causes bb_clientd to expand symlinks, returning the stat information for the  file it targets. This is somewhat complex, as Bazel (or `no-sandbox` build actions that it spawns) may create symlinks inside the FUSE file system that have absolute targets.

- `output_path_prefix` allows Bazel to specify at which absolute path the FUSE file system is placed.
- `output_path_aliases` allows Bazel to specify additional absolute path prefixes that end up resolving to locations inside the FUSE file system.

With these two pieces of knowledge, bb_clientd is capable of expanding absolute symlinks, reliably answering whether the symlink points to a location inside/outside the FUSE file system.

`output_path_aliases` is a map, where:
- the key corresponds to an absolute path **outside** of the FUSE file system
- the value corresponds to a relative path **inside** the FUSE file system.

Reply:
This is a little surprising that the FS needs to know where it is mounted to stat, technically it could just delegate to `stat` (as opposed to 'lstat') and get the info. Is that an optimization?